### PR TITLE
client: wire up ceph_ll_readv and ceph_ll_writev

### PR DESCRIPTION
**Pull request by jtlayton**
Ganesha is adding vectorized I/O operations, and samba could probably use them too. This wires them in properly.

First a cleanup: some of the underlying infrastructure was carelessly casting uint64_t values to int32_t when returning read/write lengths. That's potentially problematic if someone issues a large I/O (possibly spread over a large iovec array) as the result could end up looking negative after success. This cleans that code up and pushes the casts up into the callers.

This was necessary as ceph_ll_readv/writev return int64_t, unlike preadv/pwritev which return int. We may want to follow up here with saner handling for really large I/Os when we can't actually return a size that large. (Maybe clamp the I/O sizes in those functions to INT_MAX?)

ceph_ll_readv/writev are defined in the header, but the code always just returns -1. Wire it into the preadv/pwritev code, and simply avoid doing the fd -> Fh conversion in those codepaths.

Finally, add a testcase for ceph_ll_readv and ceph_ll_writev. This is just a clone of the ceph_preadv/ceph_pwritev test, but done with ceph_ll_readv/ceph_ll_writev.

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -9147,7 +9147,7 @@ int Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
   utime_t lat;
   uint64_t totalwritten;
   int have;
-  int r = get_caps(in, CEPH_CAP_FILE_WR|CEPH_CAP_AUTH_SHARED,
+  int64_t r = get_caps(in, CEPH_CAP_FILE_WR|CEPH_CAP_AUTH_SHARED,
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: get_caps returns a real int, not an int64.

File: src/client/Client.cc [hunk 2]
```diff
@@ -8707,7 +8707,7 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
   return _preadv_pwritev(fd, iov, iovcnt, offset, false);
 }
 
-int Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
+int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
```
Comment (gregsfortytwo) [inline src/client/Client.cc]: I'm a bit confused by what you mean about converting int64->int "at a higher level".
eg, ::_read() is called by ::read() up at line 8694, and that's stuffing the _read() result into an int. Did you just miss one, or did you mean for that to be the conversion point?

If you're worried about integer overflows, shouldn't there be size checks going in on or something? I don't really understand what it has to do with preadv/pwritev support.

File: src/client/Client.cc (hunk 1 above)
Comment (jtlayton) [reply src/client/Client.cc]: Oops, that one is probably superfluous. I'll remove that delta. Good catch.

File: src/client/Client.cc (hunk 2 above)
Comment (jtlayton) [reply src/client/Client.cc]: No, my intent was to not change how the existing functions that return an int work. They still cast the result to an int for now, but the underlying layer that ll_readv/ll_writev now call should be 64-bit clean.

...and yeah, I think we probably do want to do size checking better than we do now, but we have API functions like this:

```
int ceph_write(struct ceph_mount_info *cmount, int fd, const char *buf, int64_t size, int64_t offset);
```

What should we do when someone gives a "size" value that is larger than INT_MAX ? I think we probably want to clamp the I/O size to that, right?

Comment (batrick) [reply src/client/Client.cc]: >  I think we probably want to clamp the I/O size to INT_MAX, right?

Sounds reasonable to me.

Comment (jtlayton) [reply src/client/Client.cc]: Ok, I'll see if I can add another patch to clamp the I/O sizes to INT_MAX in those functions.

**Comment by jtlayton** [comment]
Retest this, please.

**Comment by jtlayton** [comment]
Test failures look to be unrelated to anything in this PR. Should I rebase?

**Comment by gregsfortytwo** [comment]
I wouldn't worry about bluestore issues on ARM at this time; sadly.

**Comment by jtlayton** [comment]
I fixed the descriptions and went ahead and squashed in the trivial patch too.

**Comment by jtlayton** [comment]
Thanks. I think if we're at some point in the future where we're looking at bumping the soname on libcephfs, that we might want to consider cleaning up the APIs that take length values that are larger than the return types. If we know users are going to have to recompile anyway, then we should be able to make such a change safely.

**Pull request merged**